pkg/controller/release/strategy_executor.go

```diff
@@ -0,0 +1,470 @@
+package release
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog"
+
+	shipper "github.com/bookingcom/shipper/pkg/apis/shipper/v1alpha1"
+	"github.com/bookingcom/shipper/pkg/controller"
+	shippererrors "github.com/bookingcom/shipper/pkg/errors"
+	"github.com/bookingcom/shipper/pkg/util/conditions"
+	releaseutil "github.com/bookingcom/shipper/pkg/util/release"
+)
+
+type StrategyExecutor struct {
+	curr, prev, succ *releaseInfo
+	recorder         record.EventRecorder
+	hasIncumbent     bool
+}
+
+func NewStrategyExecutor(curr, prev, succ *releaseInfo, recorder record.EventRecorder, hasIncumbent bool) (*StrategyExecutor, error) {
+	return &StrategyExecutor{
+		curr:         curr,
+		prev:         prev,
+		succ:         succ,
+		recorder:     recorder,
+		hasIncumbent: hasIncumbent,
+	}, nil
+}
+
+type PipelineContinuation bool
+
+const (
+	PipelineBreak    PipelineContinuation = false
+	PipelineContinue                      = true
+)
+
+type PipelineStep func(*StrategyExecutor, conditions.StrategyConditionsMap) (PipelineContinuation, []ExecutorResult, []ReleaseStrategyStateTransition, error)
+
+type Pipeline []PipelineStep
+
+func NewPipeline() *Pipeline {
+	return new(Pipeline)
+}
+
+func (p *Pipeline) Enqueue(step PipelineStep) {
+	*p = append(*p, step)
+}
+
+func (p *Pipeline) Process(e *StrategyExecutor, cond conditions.StrategyConditionsMap) ([]ExecutorResult, []ReleaseStrategyStateTransition, error) {
+	var res []ExecutorResult
+	var trans []ReleaseStrategyStateTransition
+	for _, step := range *p {
+		cont, stepres, steptrans, err := step(e, cond)
+		if err != nil {
+			return nil, nil, err
+		}
+		res = append(res, stepres...)
+		trans = append(trans, steptrans...)
+		if cont == PipelineBreak {
+			break
+		}
+	}
+
+	return res, trans, nil
+}
+
+func genInstallationEnforcer(curr, succ *releaseInfo) PipelineStep {
+	return func(e *StrategyExecutor, cond conditions.StrategyConditionsMap) (PipelineContinuation, []ExecutorResult, []ReleaseStrategyStateTransition, error) {
+		strategy := curr.release.Spec.Environment.Strategy
+		targetStep := curr.release.Spec.TargetStep
+		isLastStep := int(targetStep) == len(strategy.Steps)-1
+
+		if targetStep >= int32(len(strategy.Steps)) {
+			err := fmt.Errorf("no step %d in strategy for Release %q",
+				targetStep, controller.MetaKey(curr.release))
+			return PipelineBreak, nil, nil, shippererrors.NewUnrecoverableError(err)
+		}
```

Comment:
It seems like this is the only error that can happen inside the enforcers, and it's the exact same thing across all of them. Would it be possible to extract choosing the strategy step to `Execute`, and not even generate a pipeline if there's an error there? It sounds to me that it would simplify things a little bit.

Reply:
Awesome observation. Thank you so much!